mypy/checker.py

```diff
@@ -507,8 +507,16 @@ def check_overlapping_overloads(self, defn: OverloadedFuncDef) -> None:
             # decorator or if the implementation is untyped -- we gave up on the types.
             inner_type = get_proper_type(inner_type)
             if inner_type is not None and not isinstance(inner_type, AnyType):
-                assert isinstance(inner_type, CallableType)
-                impl_type = inner_type
+                if isinstance(inner_type, CallableType):
+                    impl_type = inner_type
+                elif isinstance(inner_type, Instance):
+                    inner_call = get_proper_type(
+                        find_member('__call__', inner_type, inner_type, is_operator=True)
+                    )
+                    if isinstance(inner_call, CallableType):
```

Comment:
One more problem:

```python
from typing import Callable, Union, Any, overload, Protocol

class DCall(Protocol):
    def __call__(self, arg: Union[int, str]) -> None:
        ...

class D:
    def __getattr__(self, attr: str) -> DCall:
        ...   # Will return `__call__` in runtime

def dec_d(f: Callable[..., Any]) -> D:
    return D()

@overload
def f_d(arg: int) -> None: ...
@overload
def f_d(arg: str) -> None: ...
@dec_d
def f_d(arg: Union[int, str]) -> None: ...
```

This now reports: `out/ex.py:18: error: "D" not callable`, because the return type of `__getattr__` is not `CallableType`, but `Instance`.

Reply:
Sorry, it's late for me and I might be missing something, but I think that behaviour is right? `f_d(1)` will give you `TypeError: 'D' object is not callable` at runtime

Reply:
I am going to create a new issue out of it. Sadly, I don't have the time right now to fix it.

Reply:
Sorry, it's late for me, so I might be missing something, but that seems to be the correct behaviour?

```
λ cat ex.py
from typing import Callable, Union, Any, overload, Protocol

class DCall(Protocol):
    def __call__(self, arg: Union[int, str]) -> None:
        ...

class D:
    def __getattr__(self, attr: str) -> DCall:
        if attr == "__call__":
            return lambda *a, **kw: print(a, kw)
        raise AttributeError

def dec_d(f: Callable[..., Any]) -> D:
    return D()

@overload
def f_d(arg: int) -> None: ...
@overload
def f_d(arg: str) -> None: ...
@dec_d
def f_d(arg: Union[int, str]) -> None: ...

import pytest

with pytest.raises(TypeError):
    print(D()(1))
with pytest.raises(TypeError):
    print(f_d(1))
print("correct")

λ python3 ex.py
correct

λ mypy ex.py
ex.py:20: error: "D" not callable
ex.py:26: error: "D" not callable
```